File: angular-client/src/app/setting/txmortality/txmortality-base.service.ts
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { HttpClient } from '@angular/common/http';

import { TxMortalityEntity } from '../../domain/txmortality.entity';
import { PagedResponse } from '../../domain/response/paged-response';
import { DeleteByIdCheckDto } from '../../domain/response/delete-by-id-check-dto';
import { Response } from '../../domain/response/response';

export abstract class TxMortalityBaseService {
  constructor(protected http: HttpClient) {}

  public search(criteria: any): Observable<PagedResponse> {
    return this.http.post<PagedResponse>('/api/txmortality/search', criteria);
  }

  public findById(docId: string): Observable<TxMortalityEntity> {
    return this.http.get<TxMortalityEntity>('/api/txmortality/' + docId).pipe(
      map((response: TxMortalityEntity) => {
        response.modifiedTime =
          response.modifiedTime === null
            ? null
            : new Date(response.modifiedTime);
        return response;
      })
    );
  }

  public save(
    entity: TxMortalityEntity
  ): Observable<Response<TxMortalityEntity>> {
    return this.http.post<Response<TxMortalityEntity>>(
      '/api/txmortality/',
      entity
    );
  }

  public delete(docId: string): Observable<any> {
    return this.http.delete<any>('/api/txmortality/' + docId);
  }

  public deleteByIdCheck(docId: string): Observable<DeleteByIdCheckDto> {
    return this.http.delete<any>('/api/txmortality/' + docId + '/check');
  }
}

File: couch-server/src/main/java/com/fronde/server/services/txmortality/TxMortalityBaseService.java
package com.fronde.server.services.txmortality;

import com.fronde.server.domain.TxMortalityEntity;
import com.fronde.server.domain.response.DeleteByIdCheckDTO;
import com.fronde.server.domain.response.PagedResponse;
import com.fronde.server.domain.response.Response;
import com.fronde.server.services.common.ObjectIdFactory;
import com.fronde.server.services.revision.RevisionService;
import com.fronde.server.services.validation.ValidationMessage;
import com.fronde.server.services.validation.ValidationService;
import com.fronde.server.utils.BaseEntityUtils;
import com.fronde.server.utils.CSVExportUtils;
import java.util.Arrays;
import java.util.List;
import javax.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;


@Component
public abstract class TxMortalityBaseService {

  @Autowired
  protected TxMortalityRepository repository;

  @Autowired
  protected ValidationService validationService;

  @Autowired
  protected RevisionService revisionService;

  @Autowired
  protected CSVExportUtils exportUtils;

  @Autowired
  protected ObjectIdFactory objectIdFactory;

  public TxMortalityEntity findById(String docId) {
    // using orElse instead of get so we return null if the entity is not found
    return repository.findById(docId).orElse(null);
  }

  public PagedResponse<TxMortalityEntity> search(TxMortalityCriteria criteria) {
    return repository.findByCriteria(criteria);
  }

  public Response<TxMortalityEntity> save(TxMortalityEntity entity) {
    List<ValidationMessage> messages = validationService.validateEntity(entity);
    if (messages.size() == 0) {
      if (entity.getId() == null) {
        entity.setId(objectIdFactory.create());
      }

      if (entity.getDocType() == null) {
        entity.setDocType("TxMortality");
      }

      BaseEntityUtils.setModifiedFields(entity);

      revisionService.createRevision(entity);
      entity = repository.save(entity);
    }

    Response<TxMortalityEntity> response = new Response<>(entity);
    response.setMessages(messages);
    return response;
  }

  public void deleteById(String id) {
    DeleteByIdCheckDTO deleteByIdCheckDTO = deleteByIdCheck(id);
    if (deleteByIdCheckDTO != null && deleteByIdCheckDTO.getDeleteOk()) {
      repository.deleteById(id);
    } else {
      throw new RuntimeException("Delete check failed for: " + id);
    }
  }

  public abstract DeleteByIdCheckDTO deleteByIdCheck(String docId);

  public void export(TxMortalityCriteria criteria, HttpServletResponse response) {
    List<String> header = Arrays.asList(
        "Name", "NormalBpm", "Activity BPM", "Mortality BPM", "Hours Til Mort");
    List<String> props = Arrays
        .asList("name", "normalBpm", "activityBpm", "mortalityBpm", "hoursTilMort");
    criteria.setPageSize(exportUtils.getMaxRows());
    criteria.setPageNumber(1);
    PagedResponse<TxMortalityEntity> pr = this.search(criteria);
    exportUtils.export(response, pr, header, props, "txmortality");
  }

}
